# read genotypes
# proportion success by category (axiom annotation category)
# distribution of success over SNP type (RAD, tran, pre-val)
# SNPs with annotations metabolism immune etc.
# MAF distribution (s)
# Sample quality and number of NAs and call rate

# Genome individual comparison - error rate
# Different priority categories?
# What were the failed SNPs? investigate SNP metadata eg. number of mappings etc.
# SNP distribution across genome


# Great tit
# Summary of Affy Analysis Suite category
# Comparison with DNA conc

# Samples with dish QC < 0.82 were discarded
# Samples with a call rate of < 0.95 were discarded

# Ps_Metrics and Ps_Classification functions within Affymetrix Analysis Suite were performed.

# SNPs with call rates < 0.97 discarded
# SNPs identified as containing off-target variants discarded

# Need output of Ps_metrics (APT / AAS) in order to run classification